From: SeongJae Park <sj@kernel.org>
Date: Tue, 7 Oct 2025 15:57:51 -0700
Subject: [PATCH] mm/damon/core: handle mem_cgroup_from_id() NULL return case

The cgroup might be removed while DAMON is running.  Check the NULL
return case.

Signed-off-by: SeongJae Park <sj@kernel.org>
---
 mm/damon/core.c | 6 ++++++
 1 file changed, 6 insertions(+)

diff --git a/mm/damon/core.c b/mm/damon/core.c
index cc9a3fc1b95d..3d7ead53e734 100644
--- a/mm/damon/core.c
+++ b/mm/damon/core.c
@@ -2141,6 +2141,12 @@ static unsigned long damos_get_node_memcg_used_bp(
 	rcu_read_lock();
 	memcg = mem_cgroup_from_id(goal->memcg_id);
 	rcu_read_unlock();
+	if (!memcg) {
+		if (goal->metric == DAMOS_QUOTA_NODE_MEMCG_USED_BP)
+			return 0;
+		else	/* DAMOS_QUOTA_NODE_MEMCG_FREE_BP */
+			return 10000;
+	}
 	mem_cgroup_flush_stats(memcg);
 	lruvec = mem_cgroup_lruvec(memcg, NODE_DATA(goal->nid));
 	used_pages = lruvec_page_state(lruvec, NR_ACTIVE_ANON);
-- 
2.39.5

